feat: add rule for unused anchors

amimas opened the pull request:
This change implements one of the options discussed in #395 . A new option for `anchors` rule named `forbid-unused-anchors`. If the option is enabled, yamllint will report about anchors that has been declared but is not used. The option is disabled by default as yaml specs doesn't require anchors to be used.

Rest of the options for this rule is being implemented by #420.

adrienverge commented on tests/rules/test_anchors.py:
```diff
@@ -0,0 +1,92 @@
+
+from tests.common import RuleTestCase
+
+NORMAL_ANCHOR = '''---
```
Please don't use these `NORMAL_ANCHOR`, `MISS_ANCHOR_POINTER` (I guess you've tried to reproduce the other PR), just inline the YAML code examples inside tests. All yamllint tests do this, the goal is to have tested YAML code close to the list of expected errors. Easier to read, easier to debug!

amimas replied:
You're right. I was following the other PR (#420 ). I agree having the yaml codes close to the expected error will probably be easier to read/debug.

adrienverge reviewed (commented):
Hello @amimas, thanks for contributing!

Please don't commit any editor-specific conf (remove the first commit) and more generally anything that is not related to this new option. Also squash everything into one single commit, with a meaningful commit message. You need to run linters before pushing code (I saw a lot of style errors), also you can have a look at the [CONTRIBUTING](https://github.com/adrienverge/yamllint/blob/master/CONTRIBUTING.rst) file for hints.

Then I'll test and review your pull request :+1:

amimas commented:
> Please don't commit any editor-specific conf (remove the first commit) and more generally anything that is not related to this new option.

> You need to run linters before pushing code (I saw a lot of style errors), also you can have a look at the [CONTRIBUTING](https://github.com/adrienverge/yamllint/blob/master/CONTRIBUTING.rst) file for hints.

Thanks for taking a peak at this PR. I will make sure to remove that commit next time I work on this. I think this PR should be worked on after #420 . There will be a lot of conflict. I will wait for that to be finished. Hopefully it'll be soon. That's also why I didn't run the linters; I know there's more work remaining.

> Also squash everything into one single commit, with a meaningful commit message.

Are you not able to squash merge when the PR is accepted or ready to merge? Do you prefer a single squashed commit for every push/update to the PR?

adrienverge commented:
> I think this PR should be worked on after #420 . There will be a lot of conflict. I will wait for that to be finished. Hopefully it'll be soon.

#420 was open 2 years ago and I'm not sure it will be finished soon. For this reason, I just implemented the new rule myself (see #550) so you can rebase on it and work on your new option more comfortably.

In the test cases I wrote, I added some examples of unused anchors, so you could simply reuse the same YAML snippets for your new tests.


> Are you not able to squash merge when the PR is accepted or ready to merge? Do you prefer a single squashed commit for every push/update to the PR?

Yes, please :pray:, I prefer one single commit ready to merge, as per the contribution guidelines.

amimas commented:
Hi @adrienverge - Got a little time and thought I'd take a look at this PR again, now that you've implemented couple of other options for the `anchors` rule via #550 . I could use some advice as I'm new to python.

I see the latest code is using `set` for detecting `forbid-undeclared-aliases` and `forbid-duplicated-anchors`. As you may recall, `forbid-unused-anchors` requires going through entire document and only then we can evaluate whether any of the anchors are unused or not. I thought about creating a separate `set` for all the aliases. That way the difference between anchors and aliases sets will give us the anchors that are "unused". The problem is, how to report those anchors as problem with their line and column numbers. I believe I can't use a dict/list in a set (i.e. can't store additional info about anchors in the set). And if I go with other data types, then existing options will need to be refactored.

Do you have any suggestions/reference on how to implement `forbid-unused-anchors` option?

adrienverge commented:
Hi @amimas, I'm glad you liked #550!
Sorry I don't have much time myself so I can't help you with Python, but you can inspire from other source files (especially `rules/*.py`) to create a near-perfect and ready-to-merge pull request. About the implementation, what about following the suggestion I made in our discussion at https://github.com/adrienverge/yamllint/issues/395#issuecomment-1423044402?

amimas commented:
Thanks @adrienverge for the suggestion. This is what you suggested in https://github.com/adrienverge/yamllint/issues/395#issuecomment-1423044402 for the implementation:

> So, maybe a (not perfect) solution would be to output forbid-unused-anchors errors at the end of the YAML stream (StreamEndToken -- note that there can be multiple YAML documents inside the same file). The error report anchor "foo" defined but unused wouldn't match the line where &foo is defined, but would be positioned at the end of the YAML stream.

I think that means reporting "unused anchors" doesn't need to refer to the exact line where the anchor is defined. If that's the case, then I believe the implementation is probably simple as per my earlier comment. But from usability/UX perspective, it'd be helpful if the linter could refer to the correct line.

I'll look into it a bit more and see if correct lines of the anchors can be easily captured. Otherwise, that might be the next feature update.

amimas commented:
@adrienverge - When you get a chance, please review. I managed to implement it so that unused anchor issues can report the correct line. I believe that will be helpful for end users. I didn't need a big refactor of existing implementation; just a minor change that doesn't use `set`.

I didn't squash the commits yet in case more change is needed after your review. I try to keep my commits as organized/atomic as possible. You can do squash & merge if everything looks okay or I can push a final squashed commits if you insist.

btw - should all the options in the rule be enabled in strict mode? Not sure yet where that is configured.

amimas commented:
Hi @adrienverge 👋 .  what do you think about the latest update?

adrienverge reviewed (commented):
Hello @amimas,

> I didn't squash the commits yet

I have limited time and reviewing takes much, so I do insist, please use a ready-to-merge, unified commit message on every push, with a good description that is helpful for future `git show` usage (what the rule does and why, an example of error report, etc.) Check out `git log` on this project to see what is expected.

> I managed to implement it so that unused anchor issues can report the correct line. I believe that will be helpful for end users.

This is smart!
**But**, I suspect it could cause problems to some users because the reported problems can now be interleaved / unordered, which has never been the case with yamllint. For example:
```yaml
---
- &a unused
-    k:     v  
```
will report errors in the wrong order:
```
3:5: too many spaces after hyphen (hyphens)
3:12: too many spaces after colon (colons)
3:14: trailing spaces (trailing-spaces)
2:3: found unused anchor a (anchors)
```
I propose to keep your implementation (which is cool), but be ready to change it if there are legitimate complaints (the other implementation being to print the line+column in the message string instead).
What do you think?

> just a minor change that doesn't use `set`.

Using an array (`list`) is particularly inefficient and could cause performance issues: if you can't use a `set`, please use a mapping (`dict`): `context['anchors'] = {}` with the anchor name as the key.

It will be much faster to find/access an element within it.
Also, it will reduce the diff of this PR because you can keep using `if token.value in context['anchors']`.

> btw - should all the options in the rule be enabled in strict mode? Not sure yet where that is configured.

I'm not sure what you mean: are you refering to strict mode (returning a non-zero exit code upon warnings) or the `default` set of rules? The latter is configured in the `DEFAULT` variable. In my opinion (see https://github.com/adrienverge/yamllint/pull/550#issue-1630901040): yes for `forbid-undeclared-aliases`, no for the others.

About tests cases: in #550 I tried to create one unified and reusable YAML snipped that would contain all types of errors, so it could be used for `forbid-undeclared-aliases`, `forbid-duplicated-anchors` **and** `forbid-unused-anchors`. I see that you modified it, so the last one is different from the 2 first ones: removed parts, addition of `b1`, `i1`, etc. that are not anchors nor aliases... Unless there's a good reason, let's keep a unified YAML snippet. For info, this initial snippet would report:

    2:3: found unused anchor i (anchors)
    7:3: found unused anchor s (anchors)
    14:3: found unused anchor f_s (anchors)
    16:16: found unused anchor b_m (anchors)
    19:18: found unused anchor b_m_bis (anchors)

Also it looks like the added tests don't reflect the recent commit e90e0a0, please check that.

amimas commented on yamllint/rules/anchors.py:
```diff
@@ -89,30 +112,79 @@
+        if isinstance(token, (
+                yaml.StreamStartToken,
+                yaml.DocumentStartToken,
+                yaml.DocumentEndToken)):
```
Added to all the test cases and adjusted line numbers accordingly as needed.

amimas commented on yamllint/rules/anchors.py:
```diff
@@ -24,6 +24,8 @@
 * Set ``forbid-duplicated-anchors`` to ``true`` to avoid duplications of a same
   anchor.
+* Set ``forbid-unused-anchors`` to ``true`` to avoid anchors being declared but
+  not used anywhere via alias.
```
Agreed. Updated the document as suggested.

amimas commented on yamllint/rules/anchors.py:
```diff
@@ -89,30 +112,79 @@
+            anchor = {
+                "value": token.value,
+                "line": token.start_mark.line + 1,
+                "column": token.start_mark.column + 1,
```
Makes sense. Updated to follow the project convention.

amimas commented on yamllint/rules/anchors.py:
```diff
@@ -89,30 +112,79 @@
+                "value": token.value,
+                "line": token.start_mark.line + 1,
+                "column": token.start_mark.column + 1,
+                "aliasCount": 0
```
Thanks for the suggestion. Updated as suggested.

amimas commented on yamllint/rules/anchors.py:
```diff
@@ -89,30 +112,79 @@
+        # is an alias, increment the count of alias for associated anchor.
+        # This can be used to check when end of document is reached and
+        # evaluate if an anchor is unused.
+        elif (isinstance(token, yaml.AliasToken)):
```
That's true. Updated the code.

amimas commented on yamllint/rules/anchors.py:
```diff
@@ -89,30 +112,79 @@
+        if (isinstance(next, (yaml.StreamEndToken,
+                              yaml.DocumentStartToken,
+                              yaml.DocumentEndToken)) and
+                len(context['anchors']) > 0):
```
Ahh... good catch! Thanks for the suggestion. Updated the code.

amimas commented on yamllint/rules/anchors.py:
```diff
@@ -89,30 +112,79 @@
+            for anchor in context['anchors']:
+                if anchor['value'] == token.value:
+                    anchor['aliasCount'] += 1
+                    break
```
Thanks! Using `dict` wasn't as difficult as I had initially thought. Updated the code.

amimas commented:
@adrienverge - Thanks a lot for taking the time to review and all the feedbacks and suggestions. I believe I've addressed all the concerns and adopted your suggestions. When you get a chance, please review one more time.

> I do insist, please use a ready-to-merge, unified commit message on every push, with a good description that is helpful for future git show usage (what the rule does and why, an example of error report, etc.) Check out git log on this project to see what is expected.

Sure. I've squashed all my commits and followed the style of previous commit message used for implementing other options of the anchors rule. Hope it is following your expectation now.

> I propose to keep your implementation (which is cool), but be ready to change it if there are legitimate complaints (the other implementation being to print the line+column in the message string instead).
What do you think?

Sure. Even in that case the error message for `forbid-unused-anchors` option can only be reported at the end of the document. It might look odd to have line and column numbers reported in a different format than the rest. That might not be good reporting.

> Using an array (list) is particularly inefficient and could cause performance issues: if you can't use a set, please use a mapping (dict): context['anchors'] = {} with the anchor name as the key.

Thank you for the suggestion and some examples. I managed to get it working with `dict` and I agree that it's a lot better.

> btw - should all the options in the rule be enabled in strict mode? Not sure yet where that is configured.
>
> I'm not sure what you mean: are you refering to strict mode (returning a non-zero exit code upon warnings) or the default set of rules? The latter is configured in the DEFAULT variable. In my opinion (see https://github.com/adrienverge/yamllint/pull/550#issue-1630901040): yes for forbid-undeclared-aliases, no for the others.

I was referring to the "strict mode", but I mistakenly thought that mode enables every option. I thought it's an additional "configuration" similar to the default and `relaxed`. Let's ignore that 😊 .

> About tests cases: in https://github.com/adrienverge/yamllint/pull/550 I tried to create one unified and reusable YAML snipped that would contain all types of errors, so it could be used for forbid-undeclared-aliases, forbid-duplicated-anchors and forbid-unused-anchors. I see that you modified it, so the last one is different from the 2 first ones: removed parts, addition of b1, i1, etc. that are not anchors nor aliases... Unless there's a good reason, let's keep a unified YAML snippet.

I hadn't realized that. Thanks for setting up the ground work. I've updated the test case for the new option based on your setup.

> Also it looks like the added tests don't reflect the recent commit https://github.com/adrienverge/yamllint/commit/e90e0a0eb597073e98449cb6b547343993fc2c05, please check that.

This was a newer change that was merged after I submitted the last change. I've pulled in the latest change now.

adrienverge commented on yamllint/rules/anchors.py:
```diff
@@ -113,6 +143,36 @@ def check(conf, token, prev, next, nextnext, context):
+            conf['forbid-unused-anchors']):
         if isinstance(token, yaml.AnchorToken):
-            context['anchors'].add(token.value)
+            anchor_details = {
+                "line": token.start_mark.line,
+                "column": token.start_mark.column,
+                "used": False
+            }
+            context['anchors'][token.value] = anchor_details
```
```suggestion
            context['anchors'][token.value] = {
                "line": token.start_mark.line,
                "column": token.start_mark.column,
                "used": False,
            }
```

adrienverge reviewed (commented):
Hello @amimas, thanks for this update! The commit message is great and the code looks good. I still have comments (mostly style and optimizations, nothing blocking) so the next one should be the merge one.

> It might look odd to have line and column numbers reported in a different format than the rest. That might not be good reporting.

Yes, I agree. My main fear is that some scripts somewhere on the internet rely on yamllint's output ordering, and would break with this change. We'll see!

amimas commented on yamllint/rules/anchors.py:
```diff
@@ -89,15 +112,22 @@
+                yaml.StreamStartToken,
+                yaml.DocumentStartToken,
+                yaml.DocumentEndToken)):
+            context['anchors'] = dict()
```
Yup. Updated.

amimas commented on yamllint/rules/anchors.py:
```diff
@@ -113,6 +143,36 @@ def check(conf, token, prev, next, nextnext, context):
             f'found duplicated anchor "{token.value}"')
 
-    if conf['forbid-undeclared-aliases'] or conf['forbid-duplicated-anchors']:
+    if (conf['forbid-unused-anchors']):
```
Ahh... Sorry I missed it in the last update. They're fixed now in latest commit.

amimas commented on yamllint/rules/anchors.py:
```diff
@@ -113,6 +143,36 @@ def check(conf, token, prev, next, nextnext, context):
+                        context['anchors'][anchor]['line'] + 1,
+                        context['anchors'][anchor]['column'] + 1,
+                        f"found unused anchor {anchor}"
+                    )
```
Ahh... I was trying to do something like this initially but wasn't sure how. Still learning Python 😊 . Thanks for the suggestion.

adrienverge replied:
:slightly_smiling_face:

amimas commented:
> > It might look odd to have line and column numbers reported in a different format than the rest. That might not be good reporting.
>
> Yes, I agree. My main fear is that some scripts somewhere on the internet rely on yamllint's output ordering, and would break with this change. We'll see!

I see what you mean. Let's see!

adrienverge reviewed (approved):
That's great. Thanks for contributing to yamllint @amimas!

adrienverge commented:
Ah, I noticed (too late) that the error message missed quotes around the anchor name (other messages do that):

    5:3       error    found unused anchor name not used  (anchors)

I just opened https://github.com/adrienverge/yamllint/pull/570 to fix this.

amimas commented:
> Ah, I noticed (too late) that the error message missed quotes around the anchor name (other messages do that):
> 
> ```
> 5:3       error    found unused anchor name not used  (anchors)
> ```
> 
> I just opened #570 to fix this.

Sorry I missed it. Thanks for taking care of it.

amimas commented:
@adrienverge - Just checking if this feature is available for release. Please let me know if I missed. Thanks.

adrienverge commented:
Hello @amimas, I would like to wait for https://github.com/adrienverge/yamllint/pull/572 before creating the next release. Hopefully it will be integrated soon!

adrienverge commented:
@amimas this is now released in version 1.32.0 :+1:

amimas commented:
Thanks for the update and release.
